mne/inverse_sparse/mxne_optim.py

```diff
@@ -173,6 +173,49 @@ def prox_l1(Y, alpha, n_orient):
     return Y, active_set
 
 
+def _primal_l21(M, G, X, active_set, alpha, n_orient, return_primal_only=True):
+    """Primal objective for the mixed-norm inverse problem.
+
+    See :footcite:`GramfortEtAl2012`.
+
+    Parameters
+    ----------
+    M : array, shape (n_sensors, n_times)
+        The data.
+    G : array, shape (n_sensors, n_active)
+        The gain matrix a.k.a. lead field.
+    X : array, shape (n_active, n_times)
+        Sources.
+    active_set : array of bool, shape (n_sources,)
+        Mask of active sources.
+    alpha : float
+        The regularization parameter.
+    n_orient : int
+        Number of dipoles per locations (typically 1 or 3).
+    return_primal_only : bool
+        If True, returns only the primal value. Defaults to True.
+
+    Returns
+    -------
+    p_obj : float
+        Primal objective.
+    R : array, shape (n_sensors, n_times)
+        Current residual (M - G * X).
+    nR2 : float
+        Data-fitting term.
+    GX : array, shape (n_sensors, n_times)
+        Forward prediction.
+    """
+    GX = np.dot(G[:, active_set], X)
+    R = M - GX
+    penalty = norm_l21(X, n_orient, copy=True)
+    nR2 = sum_squared(R)
+    p_obj = 0.5 * nR2 + alpha * penalty
+    if return_primal_only:
+        return p_obj
+    return p_obj, R, nR2, GX
```

Comment:
I don't understand why you need `return_primal_only` if it just takes the first element in the returned tuple (i.e., doesn't save any computation or anything

Reply:
Indeed,` _primal_l21` is called at two different locations. In `dgap_l21` where the tuple is needed, and in the acceleration piece where I only need `p_obj`. I added `return_primal_only` to avoid heavy notation like `p_obj, _, _, _ = _primal_l21(...)` .

Reply:
You don't need to make it so heavy, you can just do `_primal_l21(...)[0]` and it's much more compact than return_*_only=True`

Reply:
Ok got it! Thx!